default/google_drive.py

```diff
@@ -0,0 +1,121 @@
+import logging
+from datetime import datetime
+
+from django.http import JsonResponse
+from django.utils.translation import gettext_lazy as _
+from google.auth.transport.requests import Request as GoogleRequest
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import Flow
+from googleapiclient.discovery import build
+from googleapiclient.http import MediaFileUpload
+
+from default.data_file import DataFile
+from ocdstoucan.settings import OCDS_TOUCAN_GOOGLE_API_CREDENTIALS_FILE
+
+"""
+Code based in the documentation here: https://developers.google.com/identity/protocols/oauth2/web-server
+
+See also:
+
+* Oauth2 protocol: https://developers.google.com/identity/protocols/oauth2/
+* Credentials class docs: https://google-auth.readthedocs.io/en/latest/reference/google.oauth2.credentials.html
+* Flow class docs: https://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html
+
+"""
+
+logger = logging.getLogger(__name__)
+
+SCOPES = ['https://www.googleapis.com/auth/drive.file']
+
+google_api_messages = {
+    '_default': _('There was an issue when trying to upload the file to Google Drive, please try again later.'),
+    'access_denied': _('There was an authorization issue when saving the file to Google Drive, please try again.')
+}
+
+mime_types = {
+    '_default': 'application/zip',
+    '.csv': 'text/csv',
+    '.xlsx': 'application/vnd.google-apps.spreadsheet'
+}
+
+
+def upload_to_drive(request):
+
+    datafile = DataFile(**request.session['google_drive_file'])
+    credentials = get_credentials_from_session(request)
+
+    try:
+        if credentials:
+            if credentials.expired:
```

Comment:
If the credentials expire, it automatically restarts, then the user never enters the credentials again, it's okay?

Reply:
Yes, the credentials are refreshed. The user does not need to give explicit approval again (as long as their credentials are stored in the session).